File: src/gpu/shader.cc
#include "shader.h"

#include <format>
#include <fstream>
#include <stdexcept>
#include <vector>

#include "vulkan/vulkan_core.h"

using namespace villa;

void Shader::init(VkDevice device, const char *file_path) {
   device_ = device;

   std::ifstream file(file_path, std::ios::ate | std::ios::binary);
   file.unsetf(std::ios::skipws);

   if (!file.is_open()) {
      throw std::runtime_error(std::format("couldn't open {}", file_path));
   }

   file.seekg(0, std::ios::end);
   std::vector<char> data(file.tellg());

   file.seekg(0, std::ios::beg);
   file.read(data.data(), data.size());

   VkShaderModuleCreateInfo create_info = {
       .sType = VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO,
       .codeSize = data.size(),
       .pCode = reinterpret_cast<uint32_t *>(data.data()),
   };

   if (vkCreateShaderModule(device, &create_info, nullptr, &module_) != VK_SUCCESS) {
      throw std::runtime_error(std::format("couldn't compile shader type {}", 0));
   }
}

void Shader::deinit() {
   if (module_ != VK_NULL_HANDLE) {
      vkDestroyShaderModule(device_, module_, nullptr);
   }
}

File: runtime/gpu/vulkan/shader.zig
const std = @import("std");
const builtin = @import("builtin");
const build_options = @import("build_options");
const vk = @import("vulkan");
const vkAssert = @import("status.zig").vkAssert;

pub const Shader = struct {
    device: vk.VkDevice,
    module: vk.VkShaderModule,

    const Self = @This();

    pub fn init(device: vk.VkDevice, comptime shader_code: []const u8) Self {
        const create_info = vk.VkShaderModuleCreateInfo{
            .sType = vk.VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO,
            .codeSize = shader_code.len,
            .pCode = shader_code.ptr,
        };
        var shader_module: vk.VkShaderModule = null;
        vkAssert(vk.vkCreateShaderModule(device, &create_info, null, &shader_module)) catch
            @panic("Failed to create shader module");

        return .{ .device = device, .module = shader_module };
    }

    pub fn deinit(self: *Self) void {
        vk.vkDestroyShaderModule(self.device, self.module, null);
    }
};
